docs/root/intro/arch_overview/auxiliary_config_containers.rst

```diff
@@ -0,0 +1,66 @@
+.. _arch_overview_auxiliary_config_containers:
+
+Envoy provides various auxiliary containers for config data in it's architecture,
+to help components at different layers to work correctly.
+
+
+Per Request config container
+============================
+
+StreamInfo.dynamicMetadata
+--------------------------
+
+StreamInfo.dynamicMetadata in StreamInfo allows information to be passed along
+the filter chain. A generator filter generates the data and store it in
+dynamicMetadata by an agreed key, and a consumer filter fetches the data via the agreed key.
+
+For example, HeaderToMetadataFilter can generate metadata from headers, stores
+it with key "envoy.lb", which gets consumed later by Router to make a load
+balancing decision based on metadata match result.
+
+StreamInfo.perRequestState
+--------------------------
+
+Similar to StreamInfo.dynamicMetadata, but with more constraints,
+perReuqestState allows a filter to pass structured data (a FilterState::Object)
+to other filters, which provides a convenience interface for accessing the
+generated data. Noteworthy when requested with a non-existing key,
+an EnvoyException will be be thrown.
+
+Metadata
+========
+
+Metadata provides additional inputs to filters. It's structured as a map from
+string key, usually a filter name (in reverse DNS format), to an opaque
+goog.protobuf.Struct object. Many layers of abstractions in
+Envoy has metadata which could serve different purposes.
+
+For example, a cluster metadata may contain a per-service config entry, which
```

Comment:
Arguably we could add a per-cluster config that is distinct from the metadata. If metadata is keyed by destination filter, then this doesn't apply cleanly to a global per-cluster immutable config.